Framework/Algorithms/src/CombineDiffCal.cpp

```diff
@@ -0,0 +1,195 @@
+
+/// Algorithms name for identification. @see Algorithm::name
+const std::string CombineDiffCal::name() const { return "CombineDiffCal"; }
+
+/// Algorithm's version for identification. @see Algorithm::version
+int CombineDiffCal::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string CombineDiffCal::category() const { return "Diffraction\\Utility"; }
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string CombineDiffCal::summary() const {
+  return "Combine a per-pixel calibration with a grouped spectrum calibration";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void CombineDiffCal::init() {
+  declareProperty(
+      std::make_unique<WorkspaceProperty<DataObjects::TableWorkspace>>("PixelCalibration", "", Direction::Input),
+      "OffsetsWorkspace generated from cross-correlation. This is the source of DIFCpixel.");
+  declareProperty(
+      std::make_unique<WorkspaceProperty<DataObjects::TableWorkspace>>("GroupedCalibration", "", Direction::Input),
+      "DiffCal table generated from calibrating grouped spectra. This is the source of DIFCgroup.");
+  declareProperty(
+      std::make_unique<WorkspaceProperty<API::MatrixWorkspace>>("CalibrationWorkspace", "", Direction::Input),
+      "Workspace where conversion from d-spacing to time-of-flight for each spectrum is determined from. This is the "
+      "source of DIFCarb.");
+  declareProperty(
+      std::make_unique<WorkspaceProperty<DataObjects::TableWorkspace>>("OutputWorkspace", "", Direction::Output),
+      "DiffCal table generated from calibrating grouped spectra");
+  declareProperty(std::make_unique<WorkspaceProperty<DataObjects::MaskWorkspace>>("MaskWorkspace", "", Direction::Input,
+                                                                                  API::PropertyMode::Optional),
+                  "MaskedWorkspace for PixelCalibration");
+}
+
+int getTableWorspaceSortDirection(API::ITableWorkspace_sptr ws) {
+  Mantid::API::TableRow row = ws->getFirstRow();
+  int sortDirection = 1;
+  if (ws->rowCount() > 2) {
+    Mantid::API::TableRow lastRow = ws->getRow(ws->rowCount() - 1);
+    if (row.Int(0) > lastRow.Int(0)) {
+      sortDirection = -1;
+    }
+  }
+  return sortDirection;
+}
+
+/// sort the calibration table according increasing values in column "detid"
+API::ITableWorkspace_sptr CombineDiffCal::sortTableWorkspace(DataObjects::TableWorkspace_sptr &table) {
+  auto alg = createChildAlgorithm("SortTableWorkspace");
+  alg->setLoggingOffset(1);
+  alg->setProperty("InputWorkspace", table);
+  alg->setProperty("OutputWorkspace", table);
+  alg->setProperty("Columns", "detid");
+  alg->executeAsChildAlg();
+
+  return alg->getProperty("OutputWorkspace");
+}
+
+bool isTableWorkspaceSortedById(DataObjects::TableWorkspace_sptr ws) {
```

Comment:
And this can be removed?

Reply:
@rosswhitfield I've removed these and fixed binary search, should no longer inf loop